src/tree.rs

```diff
@@ -0,0 +1,128 @@
+use dashmap::DashMap;
+use std::{
+    any::TypeId,
+    collections::{btree_map::Entry, BTreeMap},
+    fmt::Display,
+    ops::Deref,
+    sync::{Arc, Mutex},
+};
+
+use crate::CONTAINER_TREE;
+
+struct TreeContainer<T: ?Sized> {
+    tree: Mutex<BTreeMap<Arc<T>, ()>>,
+}
+
+impl<T: Ord + ?Sized> TreeContainer<T> {
+    pub fn new() -> Self {
+        TreeContainer {
+            tree: Mutex::new(BTreeMap::new()),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct InternedTree<T: ?Sized + Ord + 'static>(Arc<T>);
+
+impl<T: ?Sized + Ord + 'static> InternedTree<T> {
+    pub fn references(this: &Self) -> usize {
+        Arc::strong_count(&this.0)
+    }
+}
+
+impl<T: ?Sized + Ord + 'static + Display> Display for InternedTree<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl<T: ?Sized + Ord + 'static> Deref for InternedTree<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
+    }
+}
+
+impl<T: ?Sized + Ord + 'static> Drop for InternedTree<T> {
+    fn drop(&mut self) {
+        if Arc::strong_count(&self.0) == 2 {
```

Comment:
This does not seem completely leak-safe to me. If the same interned value is dropped concurrently on 2 threads, potentially both `drop` invocations observe strong_count=3 in this spot.

Reply:
Can’t we just put a Weak version of the Arc into the set of interned values?
https://doc.rust-lang.org/beta/std/sync/struct.Weak.html
Granted, somehow the expired pointers must be cleaned up, but at least the values behind them will be dropped promptly when not needed anymore. Could be done just for the large interned things.

Reply:
No, I asked myself that and it can’t work, unfortunately: Weak will keep the value from being deallocated (it only breaks strong allocation cycles, so drop will happen but not dealloc), and internal mutability of a key breaks BTreeMap traversal. Will think more about the race condition — perhaps I need to bite the bullet and copy the Arc impl (what I need is a callback when strong_count reaches 1).